misk/src/main/kotlin/misk/eventrouter/RealEventRouter.kt

```diff
@@ -22,36 +22,45 @@ internal class RealEventRouter : EventRouter {
   @Inject @ForEventRouterSubscribers lateinit var subscriberExecutor: ExecutorService
 
   internal lateinit var clusterSnapshot: ClusterSnapshot
-  private val actionQueue = LinkedBlockingQueue<Action>()
-  private val localSubscribers = LinkedHashMultimap.create<String, LocalSubscription<*>>()
+  internal val hasClusterSnapshot = AtomicBoolean()
+  private val localSubscribers = LinkedHashMultimap.create<String, LocalSubscriber<*>>()
   private val remoteSubscribers = LinkedHashMultimap.create<String, WebSocket>()
-  private val hasClusterSnapshot = AtomicBoolean()
+  private val actionQueue = LinkedBlockingQueue<Action>()
+  private var hasJoinedCluster = AtomicBoolean()
 
-  // TODO(tso): Change this to LinkedHashMap and use computeIfAbsent
   private val hostToSocket = CacheBuilder.newBuilder()
       .build<String, WebSocket>(object : CacheLoader<String, WebSocket>() {
         override fun load(hostname: String): WebSocket {
-          logger.debug { "[${clusterSnapshot.self}]: connecting to $hostname" }
+          logger.info { "connecting to ${shortName(hostname)}" }
           return clusterConnector.connectSocket(hostname, webSocketListener)
         }
       })
 
   sealed class Action {
-    class OnMessage(val webSocket: WebSocket, val text: String) : Action()
-    class ClusterChanged(val newSnapshot: ClusterSnapshot) : Action()
-    class Publish(val topic: String, val event: Any) : Action()
-    class Subscribe(val localSubscription: LocalSubscription<*>) : Action()
-    class CancelSubscription(val localSubscription: LocalSubscription<*>) : Action()
+    data class OnMessage(val webSocket: WebSocket, val text: String) : Action()
```

Comment:
why are these data classes now?

Reply:
it keeps being a pain to not have nice toString functions for printing things. I can change it if it matters